libbeat/publisher/queue/diskqueue/segments.go

```diff
@@ -91,6 +95,12 @@ type queueSegment struct {
 	// A segment id is globally unique within its originating queue.
 	id segmentID
 
+	// If this segment was created during a previous session, the header
+	// field will be populated during the initial scan on queue startup.
+	// This is only used to support old schema versions, and is empty for
+	// segments created in the current session.
+	header *segmentHeader
```

Comment:
The logical dependency is that adding the frame count changed the length of the header, so it needs to know the schema version to map offsets (which are relative to the data region) to absolute disk positions. It could be used similarly if there were future changes to the header size, although we could probably remove it if / when we're confident there will be no more changes.

Reply:
I think we always need to support for migrating between versions. In the very worst case even by having two 'segment' implementation hidden behind an interface.

We didn't really change the frames format, and as I understand we do not modify existing segment files upon restart. This makes me wonder if we really need this, if we were to track file offsets by absolute positions.

Reply:
I couldn't completely remove this, but it's simplified and narrower now: segments that were loaded from a previous run have their `schemaVersion` set, which is needed to calculate the location of the first data frame. Otherwise I merged the header's frame count with the implicit one already in the segment structure (previously `framesWritten`, now `frameCount`).